autotests added

iscarrotkate opened the pull request.

iscarrotkate committed:
autotests added

AlexKitIt commented on tests/conftest.py:
```diff
@@ -0,0 +1,46 @@
+
+from tools.pages import *
+
+class User:
+
+    def __init__(self, name = None, email = None, password = None):
+        self.name = name
+        self.email = email
+        self.password = password
```
Нужно исправить: в conftest не хранится ничего кроме фикстур. Если этот класс нужен, для него нужен отельный модуль

iscarrotkate replied:
Good point, спасибо! Оставила в conftest только фикстуры, класс User заменила на простой словарь.

AlexKitIt commented on tests/conftest.py:
```diff
@@ -0,0 +1,46 @@
+
+@pytest.fixture
+def registration_page():
+    page = RegistrationPage()
```
Я очень надеюсь, что это не попытка реализовать POM тк в спринте объясняли совсем другой подход для решения этой задачи, а POM тут реализован неверно.

iscarrotkate replied:
Не знаю, что такое POM, наверное это будет дальше по курсу, а я тут изобрела велосипед. Избавилась от всего модуля, вместо обращения к атрибутам класса страницы теперь будут просто локаторы.

AlexKitIt commented on tests/test_authorization.py:
```diff
@@ -0,0 +1,81 @@
+@pytest.mark.parametrize(
+    'page, login_button',
+    [
+        ['https://stellarburgers.nomoreparties.site', sticky_bar_open_profile_button],
+        ['https://stellarburgers.nomoreparties.site', main_enter_to_account_button],
+        ['https://stellarburgers.nomoreparties.site/register', registration_login_button],
+        ['https://stellarburgers.nomoreparties.site/forgot-password', recovery_login_button]
```
Можно лучше: урлы стоит хранить отдельно, они постоянно переиспользуются, для них стоит создать модуль urls или отдельный класс в data либо модуль urls

iscarrotkate replied:
Не очень поняла про "отдельный класс в data", может тоже будет дальше, поэтому перенесла в отдельный файл по аналогии с локаторами, спасибо!

AlexKitIt commented on tests/test_authorization.py:
```diff
@@ -0,0 +1,81 @@
+        ['https://stellarburgers.nomoreparties.site/forgot-password', recovery_login_button]
+    ]
+)
+def test_open_login_form_by_unauthorized_user(authorization_page, page, login_button):
```
Нужно исправить: здесь и далее по тестам: тестовые методы в модулях объединяются в тестовые классы, необходимо создать в каждом тестовом модуле тестовые классы и разместить в них методы, даже если тестовый метод только один

AlexKitIt commented on tests/test_authorization.py:
```diff
@@ -0,0 +1,81 @@
+)
+def test_open_login_form_by_unauthorized_user(authorization_page, page, login_button):
+
+    driver = webdriver.Chrome()
```
Нужно исправить: здесь и далее: инстанс драйвера должен подниматься и закрываться в фикстуре

iscarrotkate replied:
Спасибо, добавила фикстуру. Хотя мне кажется, такого способа решения задачи (с помощью yield) не было в пройденной теории, но иначе я не придумала, как это решить.

AlexKitIt commented on tests/test_authorization.py:
```diff
@@ -0,0 +1,81 @@
+    WebDriverWait(driver, 3).until(expected_conditions.visibility_of_element_located((By.XPATH, authorization_page.header)))
+    assert driver.current_url == authorization_page.url
+
+    driver.quit()
```
Нужно исправить: logout не имеет отношения к login. Он должен быть в отдельном тестовом классе или модуле

iscarrotkate replied:
Не уверена, что согласна, но ладно, вынесла отдельно.

AlexKitIt commented on tests/test_constructor.py:
```diff
@@ -0,0 +1,25 @@
+
+
+@pytest.mark.parametrize('constructor_option', [constructor_fillings, constructor_sauces])
+def test_navigation_between_constructor_sections(main_page, constructor_option):
```
Нужно исправить: тестов на переход по табам должно быть три отдельных, по одному на каждый таб

iscarrotkate replied:
А вот тут точно не поняла, в чем проблема, можешь пояснить, зачем три теста? В принципе, согласна, что может лучше разделить на два, чтобы два лишних раза не проверять возвращение на дефолтную вкладку: 1) навигация на соусы и начинки, 2) возвращение на дефотлную вкладку, после выбора другой. Но три-то зачем?

AlexKitIt replied:
Атомарность тестов - один из основных принципов составления тестов.  Каждый упавший тест должен показывать конкретную функциональность. Правки по коду могут вносить по любому отдельному табу. Тест упал - мы знаем что конкретно сломалось а не в какой области искать проблему

iscarrotkate replied:
Просто для общего развития, я бы все же хотела продолжить обсуждение, если ты не против (а если против, я пойму игнор). Разве параметризованный тест не представляет собой по сути несколько тестов, просто удобно оформленных? В отчете они же отображаются как разные тесты и проблему локализовать не труднее, чем когда это три разных тестах. То есть, если бы не было параметризации и я в одном тесте прокликала все кнопки, то тогда это было бы недостаточно атомарно и отчет о тестировании был бы недостаточно точен т.к. с первого взгляда непонятно, какие кнопки работают, а какие нет. А с параметризацией, кажется, довольно прозрачно.

AlexKitIt commented on tests/test_constructor.py:
```diff
@@ -0,0 +1,25 @@
+
+    target_section.click()
+
+    assert 'current' in target_section.find_element(By.XPATH,"./parent::*").get_attribute("class")
```
Нужно исправить: здесь и далее: все локаторы хранятся отдельно от тестов в модуле locators, а тут только переменные, в которые они были записаны

AlexKitIt commented on tools/functions.py:
```diff
@@ -0,0 +1,44 @@
+
+from tools.pages import RegistrationPage, AuthorizationPage, MainPage
+
+def login_generator():
+
+    return 'ekaterina_baranova_19_' + str(randint(100, 999)) + '@email.com'
+
+def password_generator():
+
+    return randint(100000, 999999)
```
Под подобные методы создается модуль helpers

iscarrotkate committed:
fix tests:
1) class User removed from conftest and replaced by dictionary
2) pages entities removed from project
3) logout split to separate filr
4) all tests added to classes
5) fixed locator usage in Constructor tests
6) Constructor tests split to 3 separate tests
7) urls added to separate file
8) driver start and quit added to fixture

AlexKitIt commented on tests/conftest.py:
```diff
@@ -0,0 +1,17 @@
+from selenium import webdriver
+import pytest
+
+@pytest.fixture
+def existing_user():
+
+    return {
+        'name': 'carrot_kate_123',
+        'email': 'ek_bar_19_111@test.com',
+        'password': 'qwerty'
+    }
```
Нужно исправить: фикстуры не занимаются прокидыванием данных в тест и выполнением примитивной логики, они выполняют сложную логику предусловий\постусловий и вычислений.  Эти данные необходимо просто разместить в data и оттуда  просто импортировать куда необходимо

AlexKitIt commented on tests/test_authorization.py:
```diff
@@ -0,0 +1,70 @@
+
+        assert driver.current_url == authorization_page_url
+
+    def test_logout(self, driver, existing_user):
+
+        login(driver,existing_user)
+
+        driver.find_element(By.XPATH, sticky_bar_open_profile_button).click()
+        WebDriverWait(driver, 3).until(expected_conditions.visibility_of_element_located((By.XPATH, profile_header)))
+
+        driver.find_element(By.XPATH, profile_exit_button).click()
+
+        WebDriverWait(driver, 3).until(expected_conditions.visibility_of_element_located((By.XPATH, authorization_header)))
+        assert driver.current_url == authorization_page_url
```
А этот почему тут остался, он же теперь в TestLogout
